modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java

```diff
@@ -503,4 +503,53 @@ public static String encodeBase64(String receiver) {
     public static String decodeBase64(String receiver) {
         return new String(Base64.getDecoder().decode(receiver.getBytes(StandardCharsets.UTF_8)), StandardCharsets.UTF_8);
     }
+
+    /**
+     * Split 'receiver' by 'token' as many times as possible..
+     */
+    public static String[] splitOnToken(String receiver, String token) {
+        return splitOnToken(receiver, token, -1);
+    }
+
+    /**
+     * Split 'receiver' by 'token' up to 'limit' times.  Any limit less than 1 is ignored.
+     */
+    public static String[] splitOnToken(String receiver, String token, int limit) {
+        // Check if it's even possible to perform a split
+        if (receiver == null || token == null || receiver.length() < token.length()) {
+            return new String[] { receiver };
+        }
+
+        // List of string segments we have found
+        ArrayList<String> result = new ArrayList<String>();
+
+        // Keep track of where we are in the string
+        // indexOf(tok, startPos) is faster than creating a new search context ever loop with substring(start, end)
+        int pos = 0;
+
+        // Loop until we hit the limit or forever if we are passed in less than one (signifying no limit)
+        // If Integer.MIN_VALUE is passed in, it will still continue to loop down to 1 from MAX_VALUE
+        // This edge case should be fine as we are limited by receiver length (Integer.MAX_VALUE) even if we split at every char
+        for(;limit != 1; limit--) {
+
+            // Find the next occurrence of token after current pos
+            int idx = receiver.indexOf(token, pos);
+
+            // Reached the end of the string without another match
+            if (idx == -1) {
+                break;
+            }
+
+            // Add the found segment to the result list
+            result.add(receiver.substring(pos, idx));
+
+            // Move our search position to the next possible location
+            pos = idx + token.length();
+        }
+        // Add the remaining string to the result list
+        result.add(receiver.substring(pos));
+
+        // O(N) or faster depending on implementation
+        return result.toArray(new String[0]);
```

Comment:
If you make the ```new String[0]``` into ```new String[results.size()]``` the array will not have to be allocated twice.

Reply:
From what I have been reading, it's more efficient to use a empty array here:
https://stackoverflow.com/questions/5374311/convert-arrayliststring-to-string-array
https://shipilev.net/blog/2016/arrays-wisdom-ancients/

Reply:
Interesting articles.  Thank you!